Adopt typed throws in withUnsafe(Mutable)BufferPointer

DougGregor opened the pull request:
Fixes rdar://131405937

DougGregor committed:
Adopt typed throws in `withUnsafeBufferPointer`.

Part of rdar://131405937.

DougGregor committed:
Adopt typed throws for withUnsafeMutableBufferPointer

DougGregor committed:
Use `@_alwaysEmitIntoClient` for new entrypoints

DougGregor committed:
Add a test for typed throws and withUnsafe(Mutable)BufferPointer

DougGregor commented:
@swift-ci please test

slavapestov commented on stdlib/public/core/ArrayBuffer.swift:
```diff
@@ -567,29 +567,56 @@ extension _ArrayBuffer {
 
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @inlinable
```
This should be @usableFromInline, not @inlinable.

slavapestov commented on stdlib/public/core/Array.swift:
```diff
@@ -1595,56 +1605,24 @@ extension Array {
+  // for ABI reasons.
   @_semantics("array.withUnsafeMutableBufferPointer")
   @_effects(notEscaping self.value**)
   @inlinable // FIXME(inline-always)
```
This should be @usableFromInline, not @inlinable.

slavapestov commented on stdlib/public/core/ArrayBuffer.swift:
```diff
@@ -567,29 +567,56 @@ extension _ArrayBuffer {
   /// over the underlying contiguous storage.
   ///
   /// - Precondition: Such contiguous storage exists or the buffer is empty.
   @inlinable
```
Nit: @aeic implies @inlinable

slavapestov commented on stdlib/public/core/ArraySlice.swift:
```diff
@@ -1194,12 +1204,28 @@ extension ArraySlice {
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @_semantics("array.withUnsafeMutableBufferPointer")
+  @inlinable // FIXME(inline-always)
```
@usableFromInline not @inlinable

slavapestov commented on stdlib/public/core/ContiguousArray.swift:
```diff
@@ -1135,12 +1145,28 @@ extension ContiguousArray {
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @_semantics("array.withUnsafeMutableBufferPointer")
+  @inlinable // FIXME(inline-always)
```
Same

DougGregor commented:
Thank you, @slavapestov !

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Consistently use @usableFromInline and remove redundant @inlinables

DougGregor committed:
Enable API stability checking for standard library on arm64

This test was only enabled for x86_64.

DougGregor commented:
@swift-ci please test

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1594,59 +1604,54 @@ extension Array {
-  ///   used as the return value for the `withUnsafeMutableBufferPointer(_:)`
-  ///   method. The pointer argument is valid only for the duration of the
-  ///   method's execution.
-  /// - Returns: The return value, if any, of the `body` closure parameter.
```
This doc-comment should remain on the redone public method.

DougGregor replied:
Thank you for catching this! I'm pushing a fix now.

glessard commented on test/api-digester/stability-stdlib-source.swift:
```diff
@@ -4,7 +4,7 @@
 // This currently fails on non-Intel architectures due to no baseline being
 // available and it is not possible to filter across architectures in the
 // output.
-// XFAIL: CPU=arm64 || CPU=arm64e
```
Nice!

DougGregor committed:
Restore comment accidentally removed in typed throws adoption

DougGregor commented:
@swift-ci please test

DougGregor committed:
Genereralize standard library ABI stability test for arm64 macOS

... and now that it is fixed for the platform I'm developing odd, add the
missing entries for typed throws on withUnsafe(Mutable)BufferPointer.

DougGregor commented:
@swift-ci please test
